field_audit/field_audit.py

```diff
@@ -0,0 +1,87 @@
+import threading
+from functools import wraps
+
+from django.db import models
+
+
+__all__ = ["AlreadyAudited", "audit_fields", "audited_models", "set_request"]
+
+
+class AlreadyAudited(Exception):
+    """Class is already audited."""
+
+
+def audit_fields(*field_names):
+    """Class decorator for auditing field changes on DB model instances.
+
+    Use this on Model subclasses which need field change auditing.
+
+    :param field_names: names of fields on the model that need to be audited
+    """
+    def wrapper(cls):
+        if cls in _audited_models:
+            raise AlreadyAudited(cls)
+        if not issubclass(cls, models.Model):
+            raise ValueError(f"expected Model subclass, got: {cls}")
+        cls.__init__ = _decorate_init(cls.__init__, field_names)
+        cls.save = _decorate_db_write(cls.save, field_names)
+        cls.delete = _decorate_db_write(cls.delete, field_names)
+        _audited_models.append(cls)
+        return cls
+    if not field_names:
+        raise ValueError("at least one field name is required")
+    return wrapper
+
+
+def _decorate_init(init, field_names):
+    @wraps(init)
+    def wrapper(self, *args, **kw):
+        init(self, *args, **kw)
+        AuditEvent.attach_initial_values(field_names, self)
+    from .models import AuditEvent
+    return wrapper
+
+
+def _decorate_db_write(func, field_names):
+    @wraps(func)
+    def wrapper(self, *args, **kw):
+        # for details on using 'self._state', see:
+        # - https://docs.djangoproject.com/en/dev/ref/models/instances/#state
+        # - https://stackoverflow.com/questions/907695/
+        is_create = is_save and self._state.adding
+        object_pk = self.pk if is_delete else None
+        ret = func(self, *args, **kw)
+        AuditEvent.audit_field_changes(
+            field_names,
+            self,
+            is_create,
+            is_delete,
+            _get_request(),
+            object_pk,
+        )
+        return ret
+    is_save = func.__name__ == "save"
+    is_delete = func.__name__ == "delete"
+    if not is_save and not is_delete:
+        raise ValueError(f"invalid function for decoration: {func}")
+    from .models import AuditEvent
+    return wrapper
+
+
+def _get_request():
+    try:
+        return _thread.request
+    except AttributeError:
+        return None
+
+
+def audited_models():
+    return _audited_models.copy()
+
+
+def set_request(request):
+    _thread.request = request
```

Comment:
You probably know this and maybe got the pattern from there, but commcare-hq has a very similar utility: https://github.com/dimagi/commcare-hq/blob/579b52f1037d30c8e9184214f7278bb1d6a8b9d2/corehq/util/global_request/api.py#L6. It feels a little funny to have to do this twice on every request and add another layer of middleware on every request, but I don't think practically it would cause any problems. I guess the alternative would be to make the calling project pass in a `get_request` method in the settings or something, which feels even odder. So no actual suggestions here, but I wanted to point that out in case it made you think of something.

Reply:
Oh, I didn't know that HQ does this. This is the same pattern that the `field_history` app uses (albeit, that app does this in a more subtle way). I would actually like to get rid of this completely, but it's not immediately clear how to do it any "better". The `django.core.signals.request_started` would suffice for this app (to avoid middleware). I don't know if there are any performance benefits to alternative, my naive assumption is "probably not". Even then, I would still need to use a global variable to store the request (for auditors to reference), unless I want to attach the audit state _to_ the request itself (disclosure: I didn't investigate this solution at all). If you have more thoughts on any of my rambling here, I'm open to further conversation.

Reply:
Changed to use a `ContextVar` in 5856369.